utmatrix.h

knightvmk opened the pull request:
First work
work in progress

knightvmk committed:
utmatrix.h

First work
work in progress

alvls commented on include/utmatrix.h:
```diff
@@ -24,7 +24,14 @@ class TVector
+  {
+			if (a.Size == b.Size) return a.Size;
+			if (a.Size > b.Size) return a.Size
+			else b.Size;
```
Этот метод не скомпилируется...

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
+
+	if (s < 1 || s > MAX_VECTOR_SIZE)
+		throw s;
+	if (si <0 || s> MAX_VECTOR_SIZE)
```
Ошибка в условии

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 template <class ValType>
 TVector<ValType>::~TVector()
 {
+	delete pVector;
```
Ошибка

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 template <class ValType> // доступ
 ValType& TVector<ValType>::operator[](int pos)
 {
+	if (pos<0 || pos>MAX_VECTOR_SIZE)
```
Учтите StartIndex

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 bool TVector<ValType>::operator==(const TVector &v) const
 {
+	bool success = 1;
+	if (v.Size != Size) return 0;
```
Учесть StartIndex

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
+			if v.pVector[i] != v.pVector[i] return 0;
+		}
+	}
+	return 1;
```
Используйте логические константы

knightvmk replied:
как true и false?

alvls replied:
да

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 template <class ValType> // сравнение
 bool TVector<ValType>::operator!=(const TVector &v) const
 {
+	if v == *this
```
Упростите код в этом операторе

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
+	{
+		Size = v.Size;
+		StartIndex = v.StartIndex;
+		delete pVector;
```
Ошибка

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 template <class ValType> // присваивание
 TVector<ValType>& TVector<ValType>::operator=(const TVector &v)
 {
+	if (Size != v.Size)
```
Не хватает одной проверки

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 template <class ValType> // сложение
 TVector<ValType> TVector<ValType>::operator+(const TVector<ValType> &v)
 {
+	if (Size != v.Size)
```
Неверная реализация. См. тест на сложение векторов в качестве подсказки.

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 {
+	if (Size != v.Size)
+	{
+		TVector tmp(MaxSize(v, *this), StartIndex)
```
Аналогично сложению

alvls commented on include/utmatrix.h:
```diff
@@ -62,66 +69,188 @@ class TVector
 template <class ValType> // скалярное произведение
 ValType TVector<ValType>::operator*(const TVector<ValType> &v)
 {
+	ValType res=0;
```
Аналогично сложению

alvls commented on include/utmatrix.h:
```diff
@@ -130,9 +259,10 @@ template <class ValType>
 {
 public:
-  TMatrix(int s = 10);                           
+  TMatrix(int s);           // изначально s=10;                 
```
Верните объявление, как было

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
 template <class ValType>
 TMatrix<ValType>::TMatrix(int s): TVector<TVector<ValType> >(s)
 {
+	//Size = s;
```
Добавьте проверку s и исключение

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
 template <class ValType> // конструктор копирования
 TMatrix<ValType>::TMatrix(const TMatrix<ValType> &mt):
-  TVector<TVector<ValType> >(mt) {}
+  TVector<TVector<ValType> >(mt) 
```
Этот метод реализуется автоматически через вектор. Верните, как было

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
 TMatrix<ValType>::TMatrix(const TVector<TVector<ValType> > &mt):
-  TVector<TVector<ValType> >(mt) {}
+  TVector<TVector<ValType> >(mt) 
+{
```
Этот метод реализуется автоматически через вектор. Верните, как было

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
 template <class ValType> // сравнение
 bool TMatrix<ValType>::operator==(const TMatrix<ValType> &mt) const
 {
+	if (Size != mt.Size) return 0;
```
Используйте логические константы

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
 template <class ValType> // сравнение
 bool TMatrix<ValType>::operator!=(const TMatrix<ValType> &mt) const
 {
+	if (Size != mt.Size) return 1;
```
Реализуйте через ==

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
+		delete[] pVector;
+		Size = mt.Size;
+		StartIndex = mt.StartIndex;
+			for (register int i = 0; i < Size; i++)
```
Неверная реализация

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
 template <class ValType> // присваивание
 TMatrix<ValType>& TMatrix<ValType>::operator=(const TMatrix<ValType> &mt)
 {
+	StartIndex = mt.StartIndex;
```
Не хватает одной проверки

knightvmk replied:
Исправил ошибки, но имею проблему при Code Analysis: 
error C2668: 'TVector<ValType>::TVector': ambiguous call to overloaded function

alvls commented on include/utmatrix.h:
```diff
@@ -157,39 +287,95 @@ class TMatrix : public TVector<TVector<ValType> >
+	*this = mt;
+}
+template <class ValType>
+TMatrix<ValType>::~TMatrix()
```
Деструктор здесь не нужен.

knightvmk committed:
Some fix

have a error: error C2668: 'TVector<ValType>::TVector': ambiguous call to overloaded function

knightvmk commented:
Не могу разобраться в ошибке error C2668: 'TVector::TVector': ambiguous call to overloaded function. Появляется на стадии компиляции или code analysis из test_tmatrix, ссылаясь в самый конец файла.

alvls commented:
Завтра в терминале посмотрим

knightvmk committed:
utmatrix.h

Fix troubles in terminal class

alvls commented on include/utmatrix.h:
```diff
@@ -116,35 +109,35 @@ bool TVector<ValType>::operator==(const TVector &v) const
-		delete[] pVector;
-		pVector = 0;
-		pVector = new ValType[Size];
+		if ((Size != v.Size) && (StartIndex != v.StartIndex))
```
Перевыделять память нужно только, если размеры не совпадают. StartIndex проверять не нужно

knightvmk committed:
utmatrix.h

Fix all errors. Complete work.

alvls commented on include/utmatrix.h:
```diff
@@ -1,309 +1,308 @@
+template <class ValType> // доступ
+ValType& TVector<ValType>::operator[](int pos)
+{
+	if (pos<0 || pos>Size)
```
Неверное условие

alvls commented on include/utmatrix.h:
```diff
@@ -1,309 +1,308 @@
+{
+	if (pos<0 || pos>Size)
+		throw pos;
+	return pVector[abs(pos - StartIndex)]; //if pos<SI
```
abs использовать нельзя. pos не может быть меньше StartIndex

alvls commented on include/utmatrix.h:
```diff
@@ -1,309 +1,308 @@
+{
+	if ((s < 1) || (s > MAX_VECTOR_SIZE))
+		throw s;
+	if ((si < 0) || (si >= MAX_VECTOR_SIZE))
```
Второе условие неверно

alvls commented on include/utmatrix.h:
```diff
@@ -1,309 +1,308 @@
+		if (Size != v.Size)
+		{
+			Size = v.Size;
+			StartIndex = v.StartIndex;
```
Неверная реализация

alvls commented on include/utmatrix.h:
```diff
@@ -1,309 +1,308 @@
+TMatrix<ValType>::TMatrix(int s) : TVector<TVector<ValType> >(s)
+{
+	//Size = s;
+	if (s<1 || s>MAX_MATRIX_SIZE) throw s;
```
Левое условие лишнее

alvls commented on include/utmatrix.h:
```diff
@@ -1,309 +1,308 @@
+			pVector[i] = mt.pVector[i];
+		}
+	}
+	for (register int i = 0; i < Size; i++)
```
Неверная реализация

alvls commented:
Почему у вас постоянно весь файл обновляется? Очень неудобно, когда не видно изменений, приходится просматривать все целиком!

knightvmk committed:
Add files via upload

fix remarks

knightvmk commented:
У меня отображаются изменения при открытии файла, относительно предыдущей загрузки.

alvls commented on include/utmatrix.h:
```diff
@@ -70,7 +70,7 @@ class TVector
 TVector<ValType>::TVector(int s, int si)
 {
-	if ((s < 1) || (s > MAX_VECTOR_SIZE))
+	if ((s < 1) || (s >= MAX_VECTOR_SIZE))
```
Не то исправили. Это условие было верным. Неверно второе условие на si.

alvls commented on include/utmatrix.h:
```diff
@@ -99,9 +99,8 @@ TVector<ValType>::~TVector()
-		throw pos;
-	return pVector[abs(pos - StartIndex)]; //if pos<SI
+	if ((pos - StartIndex >= 0) && (pos - StartIndex < Size))
+		return pVector[pos - StartIndex]; else throw(pos);
```
Не пишите так условный оператор. Ветка else должна быть на отдельной строке

alvls commented on include/utmatrix.h:
```diff
@@ -132,12 +131,12 @@ TVector<ValType>& TVector<ValType>::operator=(const TVector &v)
-			Size = v.Size;
-			StartIndex = v.StartIndex;
 			delete[] pVector;
 			pVector = 0;
```
Это действие лишнее

knightvmk committed:
utmatrix re

Fix remarks 2
